src/uct/cuda/cuda_copy/cuda_copy_iface.c

```diff
@@ -469,15 +410,90 @@ static uct_iface_internal_ops_t uct_cuda_copy_iface_internal_ops = {
     .ep_is_connected       = uct_base_ep_is_connected
 };
 
+static ucs_status_t uct_cuda_copy_init_ctx_rsc(unsigned int max_cuda_events,
+                                               uct_cuda_copy_ctx_rsc_t *ctx_rsc)
+{
+    ucs_status_t status;
+    ucs_mpool_params_t mp_params;
+    ucs_memory_type_t src, dst;
+
+    ucs_mpool_params_reset(&mp_params);
+    mp_params.elem_size       = sizeof(uct_cuda_copy_event_desc_t);
+    mp_params.elems_per_chunk = 128;
+    mp_params.priv_size       = sizeof(CUcontext);
+    mp_params.max_elems       = max_cuda_events;
+    mp_params.ops             = &uct_cuda_copy_event_desc_mpool_ops;
+    mp_params.name            = "cuda_copy_event_descriptors";
+    status = ucs_mpool_init(&mp_params, &ctx_rsc->cuda_event_desc);
+    if (UCS_OK != status) {
+        ucs_error("mpool creation failed");
+        return UCS_ERR_IO_ERROR;
+    }
+
+    ucs_memory_type_for_each(src) {
+        ucs_memory_type_for_each(dst) {
+            ctx_rsc->queue_desc[src][dst].stream = NULL;
+            ucs_queue_head_init(&ctx_rsc->queue_desc[src][dst].event_queue);
+        }
+    }
+
+    ctx_rsc->short_stream = NULL;
+
+    return UCS_OK;
+}
+
+ucs_status_t
+uct_cuda_copy_iface_get_ctx_rsc(uct_cuda_copy_iface_t *iface,
+                                CUcontext cuda_ctx,
+                                uct_cuda_copy_ctx_rsc_t **ctx_rsc_p)
+{
+    unsigned long long ctx_id;
+    ucs_status_t status;
+    int ret;
+    khiter_t iter;
+    uct_cuda_copy_ctx_rsc_t *ctx_rsc;
+
+#if CUDA_VERSION >= 12000
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuCtxGetId(cuda_ctx, &ctx_id));
+    if (status != UCS_OK) {
+        goto err;
+    }
+#else
+    ctx_id = 0;
+#endif
+
+    iter = kh_put(cuda_copy_ctx_rscs, &iface->ctx_rscs, ctx_id, &ret);
```

Comment:
1. map by ctx pointer instead of id (we assume cannot get pointer of released context)
 check as assert that ctx_id is equal to cached value
2. make 1st part of this function inline and use kh_get